File: src/main/kotlin/com/linqingying/cangjie/utils/slicedMap/SlicedMapImpl.java
package com.linqingying.cangjie.utils.slicedMap;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Multimap;
import com.intellij.openapi.util.Key;
import com.intellij.util.keyFMap.KeyFMap;
import kotlin.jvm.functions.Function3;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;

public class SlicedMapImpl implements MutableSlicedMap {

    private final boolean alwaysAllowRewrite;
    String debugName = "";
    @Nullable
    private Map<Object, KeyFMap> map = null;
    private Multimap<WritableSlice<?, ?>, Object> collectiveSliceKeys = null;

    public SlicedMapImpl(boolean alwaysAllowRewrite) {
        this.alwaysAllowRewrite = alwaysAllowRewrite;
    }

    public SlicedMapImpl(boolean alwaysAllowRewrite, String name) {
        this(alwaysAllowRewrite);
        this.debugName = name;
    }

    @Override
    public String toString() {
        return debugName;
    }

    @Override
    public int getSize() {
        if (map != null) {
            return map.size();
        }
        return 0;
    }

    @Nullable
    @Override
    public <K, V> V get(@NotNull ReadOnlySlice<K, V> slice, K key) {
        KeyFMap holder = map != null ? map.get(key) : null;

        V value = holder == null ? null : holder.get(slice.getKey());

        return slice.computeValue(this, key, value, value == null);
    }

    @NotNull
    @Override
    public <K, V> Collection<K> getKeys(@NotNull WritableSlice<K, V> slice) {
        assert slice.isCollective() : "Keys are not collected for slice " + slice;

        if (collectiveSliceKeys == null) return Collections.emptyList();
        return (Collection<K>) collectiveSliceKeys.get(slice);
    }


    @Override
    public void forEach(@NotNull Function3<WritableSlice, Object, Object, Void> f) {
        if (map == null) return;
        map.forEach((key, holder) -> {
            if (holder == null) return;

            for (Key<?> sliceKey : holder.getKeys()) {
                Object value = holder.get(sliceKey);

                f.invoke(((AbstractWritableSlice) sliceKey).getSlice(), key, value);
            }
        });
    }


    @Override
    public <K, V> void put(@NotNull WritableSlice<K, V> slice, K key, V value) {
        if (!slice.check(key, value)) {
            return;
        }

        if (map == null) {
            map = new OpenAddressLinearProbingHashTable<>();
        }

        KeyFMap holder = map.get(key);
        if (holder == null) {
            holder = KeyFMap.EMPTY_MAP;
        }

        Key<V> sliceKey = slice.getKey();

        RewritePolicy rewritePolicy = slice.getRewritePolicy();
        if (!alwaysAllowRewrite && rewritePolicy.rewriteProcessingNeeded(key)) {
            V oldValue = holder.get(sliceKey);
            if (oldValue != null) {
                if (!rewritePolicy.processRewrite(slice, key, oldValue, value)) {
                    return;
                }
            }
        }

        if (slice.isCollective()) {
            if (collectiveSliceKeys == null) {
                collectiveSliceKeys = ArrayListMultimap.create();
            }

            collectiveSliceKeys.put(slice, key);
        }

        map.put(key, holder.plus(sliceKey, value));
        slice.afterPut(this, key, value);
    }

    @Override
    public void clear() {
        map = null;
        collectiveSliceKeys = null;
    }

    @NotNull
    @Override
    public <K, V> ImmutableMap<K, V> getSliceContents(@NotNull ReadOnlySlice<K, V> slice) {
        if (map == null) return ImmutableMap.of();

        ImmutableMap.Builder<K, V> builder = ImmutableMap.builder();

        map.forEach((key, holder) -> {
            V value = holder.get(slice.getKey());

            if (value != null) {
                builder.put((K) key, value);
            }
        });

        return builder.build();
    }

    /**
     * 删除某个ReadOnlySlice
     */
    public <K, V> void removeBySlice(@NotNull ReadOnlySlice<K, V> slice, K key ){
        if(map != null){
            KeyFMap holder = map.get(key);
            if (holder == null) {
                holder = KeyFMap.EMPTY_MAP;
            }


            System.out.println();
        }
    }


    @Override
    public <K> void remove(K key) {
        map.remove(key);
    }
}

File: src/main/kotlin/com/linqingying/cangjie/utils/slicedMap/SlicedMapImpl.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

//package com.linqingying.cangjie.utils.slicedMap
//
//import com.google.common.collect.ArrayListMultimap
//import com.google.common.collect.ImmutableMap
//import com.google.common.collect.Multimap
//import com.intellij.openapi.util.Key
//import com.intellij.util.keyFMap.KeyFMap
//
//
//class SlicedMapImpl(val alwaysAllowRewrite: Boolean) : MutableSlicedMap {
//    private var collectiveSliceKeys: Multimap<WritableSlice<*, *>, Any>? = null
//
//    private var map: MutableMap<*, KeyFMap?>? = null
//    override fun <K, V> get(slice: ReadOnlySlice<K, V>, key: K): V? {
//        val holder = map?.get(key)
//
//        val value = holder?.get(slice.key)
//
//        return slice.computeValue(this, key, value, value == null)
//    }
//
//    override fun <K, V> getKeys(slice: WritableSlice<K, V>): Collection<K> {
//        assert(slice.isCollective) { "Keys are not collected for slice $slice" }
//
//        if (collectiveSliceKeys == null) return emptyList()
//        return (collectiveSliceKeys!![slice] as Collection<K>)
//    }
//
//    override fun <K, V> forEach(f: (WritableSlice<K, V>, Any?, Any?) -> Unit) {
//        if (map == null) return
//        map!!.forEach { (key, holder) ->
//
//            if (holder == null) return
//            for (sliceKey in holder.keys) {
//                val value = holder[sliceKey]
//                f((sliceKey as AbstractWritableSlice<K, V>).slice, sliceKey, value)
//            }
//
//        }
//    }
//
//    override fun <K, V> put(slice: WritableSlice<K, V>, key: K, value: V) {
//        if (!slice.check(key, value)) {
//            return
//        }
//
//        if (map == null) {
//            map = OpenAddressLinearProbingHashTable<Any, KeyFMap?>()
//        }
//
//        var holder = map!![key]
//        if (holder == null) {
//            holder = KeyFMap.EMPTY_MAP
//        }
//
//        val sliceKey: Key<V> = slice.key
//
//        val rewritePolicy = slice.rewritePolicy
//        if (!alwaysAllowRewrite && rewritePolicy.rewriteProcessingNeeded(key)) {
//            val oldValue = holder!!.get(sliceKey)
//            if (oldValue != null) {
//                if (!rewritePolicy.processRewrite(slice, key, oldValue, value)) {
//                    return
//                }
//            }
//        }
//
//        if (slice.isCollective) {
//            if (collectiveSliceKeys == null) {
//                collectiveSliceKeys = ArrayListMultimap.create()
//            }
//
//            collectiveSliceKeys!!.put(slice, key)
//        }
//
//        map!!.put(key, holder!!.plus<V>(sliceKey, value))
//        slice.afterPut(this, key, value)
//    }
//
//    override fun clear() {
//        TODO("Not yet implemented")
//    }
//
//    override fun <K, V> getSliceContents(slice: ReadOnlySlice<K, V>): ImmutableMap<K, V> {
//        TODO("Not yet implemented")
//    }
//}
